### PR TITLE
[storage/qmdb] qmdb variants with partitioned indices

### PR DESCRIPTION
**Pull request by roberto-bayardo**
Adds variants of the any & current databases that use a partitioned (sharded) index.   Also consolidates many more tests and adds macros to make it easy to run the same tests across all of them.  Almost all of this change is from the test reworking.

Resolves: https://github.com/commonwarexyz/monorepo/issues/1632

<sup><a href="https://cursor.com/dashboard?tab=bugbot">Cursor Bugbot</a> reviewed your changes and found no issues for commit <u>6c7ad84</u></sup>

### CODE REVIEW

File: storage/src/qmdb/any/mod.rs [hunk 1]
```diff
@@ -242,4 +249,612 @@ pub(crate) mod test {
...
+        drop(db);
+        let db = reopen_db(context.with_label("reopened")).await;
+        assert_eq!(root, db.root());
+        assert_eq!(db.get_metadata().await.unwrap(), None);
+        assert!(db.get(&k).await.unwrap().is_none());
+
+        db.destroy().await.unwrap();
+    }
+
+    // Consolidated test that runs test_any_db_multiple_commits_delete_replayed on all 8 variants.
```
Comment (danlaine) [inline storage/src/qmdb/any/mod.rs]:
```suggestion
```
stale comment?

File: storage/src/qmdb/any/ordered/mod.rs [hunk 2]
```diff
@@ -1028,40 +1028,17 @@ where
...
-    type VariableDb = variable::Db<Context, FixedBytes<4>, Digest, Sha256, TwoCap>;
-
-    /// A type alias for a variable db with Digest keys (for generic tests).
-    type DigestVariableDb = variable::Db<Context, Digest, Digest, Sha256, TwoCap>;
-
-    /// Helper trait for testing Any databases that cycle through all four states.
-    trait TestableAnyDb<V>:
+    /// Helper trait for testing Any databases with FixedBytes<4> keys.
+    /// Used for edge case tests that require specific key patterns.
+    pub(crate) trait TestableAnyDb<V>:
```
Comment (danlaine) [inline storage/src/qmdb/any/ordered/mod.rs]: We can combine this with `DigestTestableAnyDb` if we add another generic type parameter `K` and do `CleanAny<Key = K>`

File: storage/src/qmdb/current/ordered/fixed.rs [hunk 3]
```diff
@@ -102,6 +113,125 @@ impl<
...
+    use commonware_runtime::{Clock, Metrics, Storage as RStorage};
+    use commonware_utils::Array;
+
+    /// A partitioned variant of [super::Db].
+    ///
+    /// The const generic `P` specifies the number of prefix bytes used for partitioning:
+    /// - `P = 1`: 256 partitions
+    /// - `P = 2`: 65,536 partitions
+    /// - `P = 3`: ~16 million partitions
+    pub type Db<
```
Comment (danlaine) [inline storage/src/qmdb/current/ordered/fixed.rs:81]: Should we assert at compile time that `P` is in `[1, 3]` or whatever we think is a reasonable upper bound?

Comment (roberto-bayardo) [reply storage/src/qmdb/current/ordered/fixed.rs:81]: this is already done in the Index code

File: storage/src/qmdb/any/ordered/mod.rs (hunk 2 above)
Comment (roberto-bayardo) [reply storage/src/qmdb/any/ordered/mod.rs]: I'd still want concrete types for Digest vs Fixed keys so this only seems to add code?

File: storage/src/qmdb/any/mod.rs (hunk 1 above)
Comment (roberto-bayardo) [reply storage/src/qmdb/any/mod.rs]: yup removed.

File: storage/src/qmdb/any/ordered/mod.rs (hunk 2 above)
Comment (danlaine) [reply storage/src/qmdb/any/ordered/mod.rs]: This diff is what I mean. 

```
 1 file changed, 12 insertions(+), 23 deletions(-)
```

```
diff --git a/storage/src/qmdb/any/ordered/mod.rs b/storage/src/qmdb/any/ordered/mod.rs
index fc1b52647..50757b163 100644
--- a/storage/src/qmdb/any/ordered/mod.rs
+++ b/storage/src/qmdb/any/ordered/mod.rs
@@ -1048,34 +1048,21 @@ mod test {
     use commonware_utils::sequence::FixedBytes;
     use core::{future::Future, pin::Pin};
 
-    /// Helper trait for testing Any databases with FixedBytes<4> keys.
-    /// Used for edge case tests that require specific key patterns.
-    pub(crate) trait TestableAnyDb<V>:
-        CleanAny<Key = FixedBytes<4>> + MerkleizedStore<Value = V, Digest = Digest>
+    /// Helper trait for testing Any databases.
+    pub(crate) trait TestableAnyDb<K, V>:
+        CleanAny<Key = K> + MerkleizedStore<Value = V, Digest = Digest>
     {
     }
 
-    impl<T, V> TestableAnyDb<V> for T where
-        T: CleanAny<Key = FixedBytes<4>> + MerkleizedStore<Value = V, Digest = Digest>
-    {
-    }
-
-    /// Helper trait for testing Any databases with Digest keys.
-    /// Used for generic tests that can be shared with partitioned variants.
-    pub(crate) trait DigestTestableAnyDb<V>:
-        CleanAny<Key = Digest> + MerkleizedStore<Value = V, Digest = Digest>
-    {
-    }
-
-    impl<T, V> DigestTestableAnyDb<V> for T where
-        T: CleanAny<Key = Digest> + MerkleizedStore<Value = V, Digest = Digest>
+    impl<T, K, V> TestableAnyDb<K, V> for T where
+        T: CleanAny<Key = K> + MerkleizedStore<Value = V, Digest = Digest>
     {
     }
 
     /// Test an empty database with Digest keys.
     ///
     /// This function is pub(crate) so partitioned variants can call it.
-    pub(crate) async fn test_digest_ordered_any_db_empty<D: DigestTestableAnyDb<Digest>>(
+    pub(crate) async fn test_digest_ordered_any_db_empty<D: TestableAnyDb<Digest, Digest>>(
         context: Context,
         mut db: D,
         reopen_db: impl Fn(Context) -> Pin<Box<dyn Future<Output = D> + Send>>,
@@ -1127,7 +1114,7 @@ mod test {
     /// Test basic CRUD and commit behavior with Digest keys.
     ///
     /// This function is pub(crate) so partitioned variants can call it.
-    pub(crate) async fn test_digest_ordered_any_db_basic<D: DigestTestableAnyDb<Digest>>(
+    pub(crate) async fn test_digest_ordered_any_db_basic<D: TestableAnyDb<Digest, Digest>>(
         context: Context,
         db: D,
         reopen_db: impl Fn(Context) -> Pin<Box<dyn Future<Output = D> + Send>>,
@@ -1255,7 +1242,7 @@ mod test {
         db.destroy().await.unwrap();
     }
 
-    pub(crate) async fn test_ordered_any_db_empty<D: TestableAnyDb<Digest>>(
+    pub(crate) async fn test_ordered_any_db_empty<D: TestableAnyDb<FixedBytes<4>, Digest>>(
         context: Context,
         mut db: D,
         reopen_db: impl Fn(Context) -> Pin<Box<dyn Future<Output = D> + Send>>,
@@ -1307,7 +1294,7 @@ mod test {
         db.into_merkleized().await.unwrap().destroy().await.unwrap();
     }
 
-    pub(crate) async fn test_ordered_any_db_basic<D: TestableAnyDb<Digest>>(
+    pub(crate) async fn test_ordered_any_db_basic<D: TestableAnyDb<FixedBytes<4>, Digest>>(
         context: Context,
         db: D,
         reopen_db: impl Fn(Context) -> Pin<Box<dyn Future<Output = D> + Send>>,
@@ -1419,7 +1406,9 @@ mod test {
 
     /// Builds a db with colliding keys to make sure the "cycle around when there are translated
     /// key collisions" edge case is exercised.
-    pub(crate) async fn test_ordered_any_update_collision_edge_case<D: TestableAnyDb<Digest>>(
+    pub(crate) async fn test_ordered_any_update_collision_edge_case<
+        D: TestableAnyDb<FixedBytes<4>, Digest>,
+    >(
         db: D,
     ) {
         // This DB uses a TwoCap so we use equivalent two byte prefixes for each key to ensure

```

Comment (roberto-bayardo) [reply storage/src/qmdb/any/ordered/mod.rs]: Really we only need 2, one with Digest keys and one with FixedByte keys, so no reason for all this genericity. I have changed it to just directly define these two types without requiring generics. PTAL?

File: storage/src/qmdb/current/ordered/db.rs [hunk 4]
```diff
@@ -321,10 +318,10 @@ impl<
         C: MutableContiguous<Item = Operation<K, V>>,
         K: Array,
         V: ValueEncoding,
+        I: OrderedIndex<Value = Location> + 'static,
```
Comment (danlaine) [inline storage/src/qmdb/current/ordered/db.rs:321]: Do we need all the `+ 'static` in this PR? It seems to build/lint without these.

File: storage/src/qmdb/current/mod.rs [hunk 5]
```diff
@@ -608,4 +820,216 @@ pub mod tests {
             }
         });
     }
+
+    // ============================================================
+    // Consolidated tests for all 8 Current QMDB variants
```
Comment (danlaine) [inline storage/src/qmdb/current/mod.rs]:
```suggestion
    // Consolidated tests for all 12 Current QMDB variants
```

File: storage/src/qmdb/any/unordered/fixed.rs [hunk 6]
```diff
@@ -693,4 +587,84 @@ pub(crate) mod test {
...
+            test_any_db_basic(db, move |idx| {
+                let ctx = ctx.with_label(&format!("db_{}", idx + 1));
+                Box::pin(open_partitioned_db_p1(ctx))
+            })
+            .await;
+        });
+    }
+
+    #[test_traced("INFO")]
+    fn test_partitioned_p1_empty() {
```
Comment (danlaine) [inline storage/src/qmdb/any/unordered/fixed.rs:658]: Do we also want to implement these for `p2`?

File: storage/src/qmdb/current/ordered/db.rs (hunk 4 above)
Comment (roberto-bayardo) [reply storage/src/qmdb/current/ordered/db.rs:321]: yes it's needed for the older compiler required by msrv

File: storage/src/qmdb/any/unordered/fixed.rs (hunk 6 above)
Comment (roberto-bayardo) [reply storage/src/qmdb/any/unordered/fixed.rs:658]: I think these tests can be further consolidated into a version we can apply to all 12 variants with the macros, but this diff was already getting pretty big, so I just restricted this to be just on p1 for now. I did the same in any/orderded/*  (test only on p=1).

File: storage/src/qmdb/current/mod.rs (hunk 5 above)
Comment (roberto-bayardo) [reply storage/src/qmdb/current/mod.rs]: fixed

**Review by danlaine — APPROVED** [review]
